### PR TITLE
feat: add ability to override default interaction behavior

### LINKED ISSUE #412: Create a mechanism to allow Giraffe consumers to specify click callbacks
### Background
Giraffe currently handles double clicks by resetting the domain (i.e. unzooming). This is the default behavior, but this should be able to be overridden by an app that consumes Giraffe (e.g. ui)

Current mouse behavior that should be overridden:
- double click
- dragging

### Acceptance Criteria
- The consumer of Giraffe can pass in an optional callback to use during clicks to override the default double click callback

### PR DESCRIPTION
**Pull request by hoorayimhelping**
Closes #412

Adds (yet another) property to `Config`: 

```ts
interactionHandlers: InteractionHandler
```

`InteractionHandler` currently defines a single interaction, `doubleClick`. The current default behavior for doubleClick is to `resetDomains` which effectively unzooms the screen. When a consumer of `Plot` provides an `interactionHandler` callback for `doubleClick`, Giraffe sends back the following data:

`hoverX`: the x coordinate of the cursor when the double click happens
`hoverY`: the y coordinate of the cursor when the double click happens
`xDomain`: the x domain of the graph (i.e. the visible horizontal portion) when the double click happens
`yDomain`: the y domain of the graph (i.e. the visible vertical portion) when the double click happens
`resetDomains`: a function to reset the zoom back to default

---

### An example of it working in UI:

This interaction handler was added to `XYPlot`
```ts
const interactionHandlers = {
  doubleClick: (datBoi) => {
    console.log("it's dat boi'", datBoi)
    datBoi.resetDomains()
  }
}
```
and this is the result when double clicking:
Screen Shot 2020-12-10 at 11 57 33 AM

### CODE REVIEW

File: giraffe/jest.config.js [hunk 1]
```diff
@@ -1,11 +1,17 @@
 module.exports = {
   preset: 'ts-jest',
-  testEnvironment: 'node',
+  testEnvironment: 'jest-environment-jsdom-sixteen',
```
Comment (hoorayimhelping) [inline giraffe/jest.config.js:3]: this'll let us test `canvas` functionality

File: giraffe/package.json [hunk 2]
```diff
@@ -51,6 +52,7 @@
     "@typescript-eslint/eslint-plugin": "^2.10.0",
     "@typescript-eslint/parser": "^2.10.0",
     "awesome-typescript-loader": "^5.2.1",
+    "canvas": "^2.6.1",
```
Comment (hoorayimhelping) [inline giraffe/package.json:55]: needed for canvas to work with jest

File: giraffe/src/components/SizedPlot.tsx [hunk 3]
```diff
@@ -2,15 +2,15 @@ import React, {useCallback, FunctionComponent, CSSProperties} from 'react'
...
-  ScatterLayerConfig,
-  RectLayerConfig,
-  MosaicLayerConfig,
   LayerTypes,
+  LineLayerConfig,
+  MosaicLayerConfig,
+  RectLayerConfig,
+  ScatterLayerConfig,
+  SingleStatLayerConfig,
+  SizedConfig,
```
Comment (hoorayimhelping) [inline giraffe/src/components/SizedPlot.tsx:13]: alphabetizing

File: giraffe/src/utils/PlotEnv.ts [hunk 4]
```diff
@@ -397,7 +397,7 @@ export class PlotEnv {
     }
   }
 
-  public resetDomains = (): void => {
+  public resetDomains(): void {
```
Comment (hoorayimhelping) [inline giraffe/src/utils/PlotEnv.ts:400]: defining it as an instance property (as was done before) makes it not part of the public prototype interface, which makes it impossible to `spyOn`

**Comment by TCL735** [comment]
Looks good to me

**Review by TCL735 — APPROVED** [review]

**Comment by T0-B1** [comment]
Would be nice to have <code>InteractionHandlers</code> documented in the quick start guide.

**Pull request merged**